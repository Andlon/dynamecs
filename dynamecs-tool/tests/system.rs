use std::error::Error;
use escargot::CargoBuild;
use tempfile::tempdir;

#[test]
fn test_timing_basic_app1_all_steps() -> Result<(), Box<dyn Error>> {
    // We use a fixed logfile to avoid the problem of having to account for
    // varying durations. To re-generate a log file, run from the project root:
    //  $ cd test_apps
    //  $ cargo run --bin basic_app1
    //  $ cp output/logs/dynamecs_log.jsonlog ../dynamecs-tool/tests/test_logs/

    let temp_dir = tempdir()?;
    let target_dir = temp_dir.path().join("target");

    // Then run dynamecs-tool
    let output = CargoBuild::new()
        .bin("dynamecs-tool")
        .target_dir(target_dir)
        .run()?
        .command()
        .arg("timing")
        .args(["--logfile", "tests/test_logs/dynamecs_app.jsonlog"])
        .output()?;

    assert!(output.status.success());
    assert!(output.stderr.is_empty());

    let stdout_str = String::from_utf8(output.stdout)?;
    insta::assert_snapshot!(stdout_str);

    Ok(())
}