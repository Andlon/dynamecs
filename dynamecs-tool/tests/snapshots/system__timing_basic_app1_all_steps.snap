---
source: dynamecs-tool/tests/system.rs
expression: stdout_str
---
Timings for step index 0
════════════════════════════════
  Total      Average    Count  Rel parent  Rel root  Span                          
  ══════════════════════════════════════════════════════════════════════════════════
   808.0 μs   808.0 μs      1         N/A   100.0 %  step                          
    78.0 μs    39.0 μs      2       9.7 %     9.7 %  ├── post_systems              
    37.0 μs    37.0 μs      1       4.6 %     4.6 %  ├── pre_systems               
   295.0 μs   295.0 μs      1      36.5 %    36.5 %  └── simulation_systems        
   134.0 μs   134.0 μs      1      45.4 %    16.6 %      └── span1                 
    42.0 μs    42.0 μs      1      31.3 %     5.2 %          └── span2             
  ══════════════════════════════════════════════════════════════════════════════════


Timings for step index 1
════════════════════════════════
  Total      Average    Count  Rel parent  Rel root  Span                          
  ══════════════════════════════════════════════════════════════════════════════════
   624.0 μs   624.0 μs      1         N/A   100.0 %  step                          
    37.0 μs    37.0 μs      1       5.9 %     5.9 %  ├── post_systems              
    37.0 μs    37.0 μs      1       5.9 %     5.9 %  ├── pre_systems               
   281.0 μs   281.0 μs      1      45.0 %    45.0 %  └── simulation_systems        
   139.0 μs   139.0 μs      1      49.5 %    22.3 %      └── span1                 
    42.0 μs    42.0 μs      1      30.2 %     6.7 %          └── span2             
  ══════════════════════════════════════════════════════════════════════════════════


Aggregate timings
════════════════════════════════

  Total      Average    Count  Rel parent  Rel root  Span                              
  ══════════════════════════════════════════════════════════════════════════════════════
     1.8 ms     1.8 ms      1         N/A   100.0 %  run                               
     1.4 ms   716.0 μs      2      81.1 %    81.1 %  └── step                          
   115.0 μs    38.3 μs      3       8.0 %     6.5 %      ├── post_systems              
    74.0 μs    37.0 μs      2       5.2 %     4.2 %      ├── pre_systems               
   576.0 μs   288.0 μs      2      40.2 %    32.6 %      └── simulation_systems        
   273.0 μs   136.5 μs      2      47.4 %    15.5 %          └── span1                 
    84.0 μs    42.0 μs      2      30.8 %     4.8 %              └── span2             
  ══════════════════════════════════════════════════════════════════════════════════════


Number of completed time steps: 2

